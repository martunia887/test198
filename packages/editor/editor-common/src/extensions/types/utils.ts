import { ExtensionNode } from './extension-manifest';

export type MenuItem = {
  key: string;
  title: string;
  description: string;
  icon: () => Promise<any>;
  node: ExtensionNode;
};

export type MenuItemMap = { [key: string]: MenuItem };
