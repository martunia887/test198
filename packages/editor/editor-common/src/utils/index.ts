export * from './color';
export * from './macro';
export * from './validator';
export { default as browser } from './browser';
export {
  default as ErrorReporter,
  ErrorReportingHandler,
} from './error-reporter';
export * from './date';
export * from './imageLoader';
export * from './calc-breakout-width';
export { default as ADFTraversor } from './traversor';
export * from './analytics';
export { measureRender } from './performance/measure-render';
export { startMeasure, stopMeasure, clearMeasure } from './performance/measure';
export { getResponseEndTime } from './performance/navigation';
