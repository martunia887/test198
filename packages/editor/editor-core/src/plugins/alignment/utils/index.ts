import { findParentNodeOfType } from 'prosemirror-utils';
import { CellSelection } from 'prosemirror-tables';
import { AlignmentState } from '../pm-plugins/main';

export const getActiveAlignment = (state): AlignmentState | undefined => {
  if (state.selection instanceof CellSelection) {
    let marks: string[] = [];
    state.selection.forEachCell(cell => {
      const mark = cell.firstChild!.marks.filter(
        mark => mark.type === state.schema.marks.alignment,
      )[0];
      if (mark) {
        marks.push(mark.attrs.align);
      } else {
        marks.push('start');
      }
    });

    let seen;
    for (let i = 0, len = marks.length; i < len; i++) {
      const mark = marks[i];

      if (typeof seen === 'undefined') {
        seen = mark;
      } else if (seen !== mark) {
        seen = null;
        break;
      }
    }
    return seen ? seen : 'start';
  }

  const node = findParentNodeOfType([
    state.schema.nodes.paragraph,
    state.schema.nodes.heading,
  ])(state.selection);
  const getMark =
    node &&
    node.node.marks.filter(
      mark => mark.type === state.schema.marks.alignment,
    )[0];

  return (getMark && getMark.attrs.align) || 'start';
};
