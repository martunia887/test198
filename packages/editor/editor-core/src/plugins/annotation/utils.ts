import { Node } from 'prosemirror-model';
import { EditorState, TextSelection, Transaction } from 'prosemirror-state';

export function canAnnotationBeCreatedInRange(from: number, to: number) {
  return (state: EditorState) => {
    const {
      schema: {
        marks: { annotation: annotationMark },
      },
    } = state;

    if (!state.doc.rangeHasMark(from, to, annotationMark)) {
      const $from = state.doc.resolve(from);
      const $to = state.doc.resolve(to);

      if (
        $from.parent === $to.parent &&
        $from.parent.isTextblock &&
        $from.parent.type.allowsMarkType(annotationMark)
      ) {
        let allowed = true;
        state.doc.nodesBetween(from, to, node => {
          allowed =
            allowed && !node.marks.some(m => m.type.excludes(annotationMark));
          return allowed;
        });
        return allowed;
      }
    }

    return false;
  };
}

export function isSelectionInsideAnnotation(
  stateOrTr: EditorState | Transaction,
) {
  return !!stateOrTr.doc.type.schema.marks.annotation.isInSet(
    stateOrTr.selection.$from.marks(),
  );
}

export function isSelectionAroundAnnotation(
  stateOrTr: EditorState | Transaction,
) {
  const {
    selection: { $from, $to },
  } = stateOrTr;
  const node = $from.nodeAfter;

  return (
    !!node &&
    $from.textOffset === 0 &&
    $to.pos - $from.pos === node.nodeSize &&
    !!stateOrTr.doc.type.schema.marks.annotation.isInSet(node.marks)
  );
}

export function getActiveAnnotationMark(
  state: EditorState | Transaction,
): { node: Node; pos: number } | undefined {
  const {
    selection: { $from },
  } = state;

  if (
    isSelectionInsideAnnotation(state) ||
    isSelectionAroundAnnotation(state)
  ) {
    const pos = $from.pos - $from.textOffset;
    const node = state.doc.nodeAt(pos) as Node;
    return node && node.isText ? { node, pos } : undefined;
  }

  return undefined;
}

export function getActiveText(state: EditorState): string | undefined {
  const { selection, tr } = state;

  if (selection instanceof TextSelection) {
    const { $from, from, to } = selection;
    const node = tr.doc.nodeAt(from);

    if (node && node.isText) {
      return node.textContent.substr($from.textOffset, to - from);
    }
  }

  return;
}
