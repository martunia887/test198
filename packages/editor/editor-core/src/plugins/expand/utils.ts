import { EditorState } from 'prosemirror-state';
import {
  findSelectedNodeOfType,
  findParentNodeOfType,
} from 'prosemirror-utils';

export const findExpand = (state: EditorState) => {
  const { expand, nestedExpand } = state.schema.nodes;
  return (
    findSelectedNodeOfType([expand, nestedExpand])(state.selection) ||
    findParentNodeOfType([expand, nestedExpand])(state.selection)
  );
};
