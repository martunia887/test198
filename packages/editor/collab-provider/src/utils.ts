import { AllSelection, NodeSelection, Selection } from 'prosemirror-state';

import { SendableSelection } from './types';

function isAllSelection(selection: Selection) {
  return selection instanceof AllSelection;
}

function isNodeSelection(selection: Selection) {
  return selection instanceof NodeSelection;
}

export const getSendableSelection = (
  selection: Selection,
): SendableSelection => {
  /**
   * <kbd>CMD + A</kbd> triggers a AllSelection
   * <kbd>escape</kbd> triggers a NodeSelection
   */
  return {
    type: 'textSelection',
    anchor: selection.anchor,
    head:
      isAllSelection(selection) || isNodeSelection(selection)
        ? selection.head - 1
        : selection.head,
  };
};

const participants = [
  'Awilda Mcentire',
  'Katie Iddings',
  'Serafina Mcdowell',
  'Eun Knopp',
  'Walker Reighard',
  'Keely Muth',
  'Allie April',
  'Rosamond Paton',
  'Lashawnda Peppard',
  'Hubert Feingold',
  'Idalia Mcquaig',
  'Wyatt Byer',
  'Genevieve Reeder',
  'Dona Elton',
  'Jerilyn Holte',
  'Salina Erhardt',
  'Velda Latour',
  'Regina Richer',
  'Natisha Tiger',
  'Glynda Stuckey',
  'Wilson Woodberry',
  'Elvie Furtado',
  'Lasandra Palmatier',
  'Louise Mijares',
  'Herma Vanleer',
  'Titus Plante',
  'James Fiscus',
  'Ina Ramero',
  'Jeanene Bernabe',
  'Lida Lepine',
  'Rosalina Palka',
  'Dayle Grossi',
  'Zack Leister',
  'Isobel Fay',
  'Elouise Kluge',
  'Twyla Machin',
  'Diane Herrington',
  'Fumiko Elders',
  'Kennith Howe',
  'Doloris Mccollom',
  'Daryl Wherry',
  'Sade Albury',
  'Roy Cork',
  'Hunter Hultgren',
  'Mi Miley',
  'Isabelle Halperin',
  'Kayla Starke',
  'Allyson Woomer',
  'Sirena Politte',
  'Patricia Scovil',
];

function hashCode(s: string) {
  let h = 0;
  for (let i = 0; i < s.length; i++)
    h = (Math.imul(31, h) + s.charCodeAt(i)) | 0;

  return Math.abs(h);
}

export const getParticipant = (userId: string) => {
  // const id = +userId.split('/')[1];
  // console.log({userId});
  const id = hashCode(userId);
  // eslint-disable-next-line no-bitwise
  const name = participants[(id | 0) % participants.length];
  return {
    userId,
    name,
    avatar: `https://api.adorable.io/avatars/80/${name.replace(/\s/g, '')}.png`,
    email: `${name.replace(/\s/g, '').toLocaleLowerCase()}@atlassian.com`,
  };
};

export const createLogger = (prefix: string, color: string = 'blue') => (
  msg: string,
  data: any = null,
) => {
  // eslint-disable-next-line no-console
  console.log(`%c${prefix}: ${msg}`, `color: ${color}; font-weight: bold`);
  if (data) {
    // eslint-disable-next-line no-console
    console.log(data);
  }
};
