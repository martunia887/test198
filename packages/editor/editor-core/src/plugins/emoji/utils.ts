import { EditorState, StateField, PluginKey } from 'prosemirror-state';
import { Command } from '../../types';
import { Dispatch } from '../../event-dispatcher';

export type Reducer<S, A> = (state: S, action: A) => S;

export function pluginStateFactory<S, A, IS extends S>(
  pluginKey: PluginKey,
  initialState: IS,
  reducer: Reducer<S, A>,
): {
  createState: (dispatch: Dispatch) => StateField<S>;
  createCommand: (action: A) => Command;
  getPluginState: (editorState: EditorState) => S;
} {
  return {
    createState: dispatch => ({
      init: () => initialState,
      apply(tr, pluginState) {
        const meta = tr.getMeta(pluginKey);
        if (meta) {
          const newState = reducer(pluginState, meta);
          dispatch(pluginKey, newState);
          return newState;
        }
        return pluginState;
      },
    }),
    createCommand: (action): Command => (state, dispatch) => {
      dispatch(state.tr.setMeta(pluginKey, action));
      return true;
    },
    getPluginState: editorState => pluginKey.getState(editorState),
  };
}
