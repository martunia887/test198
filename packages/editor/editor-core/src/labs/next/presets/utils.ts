import { EditorPresetProps, PluginsPreset } from './types';
import { EditorPlugin } from '../../../types';
import { compose } from '../../../utils';

export const removeExcludes = (excludes: EditorPresetProps['excludes']) => (
  plugins: EditorPlugin[],
) => {
  if (excludes) {
    return plugins.filter(plugin => excludes.indexOf(plugin.name) === -1);
  }
  return plugins;
};

export type ExperimentalPluginMap = Map<string, EditorPlugin>;
export const enableExperimental = (
  experimental: EditorPresetProps['experimental'],
  experimentalPluginMap: ExperimentalPluginMap,
) => (plugins: EditorPlugin[]) => {
  if (experimental && experimental.length) {
    experimental.map(pluginName => {
      const plugin = experimentalPluginMap.get(pluginName);
      if (plugin) {
        plugins.push(plugin);
      }
    });
  }
  return plugins;
};

export const processPluginPreset = (preset: PluginsPreset) => {
  const cache = new Map();
  preset.forEach(pluginEntry => {
    if (Array.isArray(pluginEntry)) {
      const [fn, options] = pluginEntry;
      cache.set(fn, options);
    } else {
      /**
       * Prevent usage of same plugin twice without override.
       * [
       *  plugin1,
       *  [plugin1, { option1: value }],
       *  plugin1, // This will throw
       * ]
       */
      if (cache.has(pluginEntry) && cache.get(pluginEntry) === undefined) {
        throw new Error(`${pluginEntry} is already included!`);
      }
      cache.set(pluginEntry, undefined);
    }
  });

  let plugins: Array<EditorPlugin> = [];
  cache.forEach((options, fn) => {
    plugins.push(fn(options));
  });

  return plugins;
};

export const getPluginsFromPreset = (
  preset: PluginsPreset,
  excludes?: EditorPresetProps['excludes'],
  experimental?: EditorPresetProps['experimental'],
): Array<EditorPlugin> => {
  const experimentalMap: ExperimentalPluginMap = new Map();

  return compose(
    enableExperimental(experimental, experimentalMap),
    removeExcludes(excludes),
  )(processPluginPreset(preset));
};
