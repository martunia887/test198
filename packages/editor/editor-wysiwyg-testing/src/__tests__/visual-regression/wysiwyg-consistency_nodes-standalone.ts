import { createDocumentADF } from './utils/_adf-utils';
import { loadKitchenSinkWithAdf } from './utils/_example-utils';
import { snapshotAndCompare } from './utils/_comparison-utils';
import { standaloneNodes } from './_data';

describe('WYSIWYG Snapshot Test: Standalone nodes look consistent in editor & renderer', () => {
  let page: any;

  beforeAll(async () => {
    // @ts-ignore
    page = global.page;
    await page.setViewport({ width: 2000, height: 1000 });
  });

  it.each(standaloneNodes)('%p', async ({ node, waitFor }) => {
    const adf = createDocumentADF(node);
    await loadKitchenSinkWithAdf(page, adf);
    await snapshotAndCompare(page, node, waitFor);
  });
});
