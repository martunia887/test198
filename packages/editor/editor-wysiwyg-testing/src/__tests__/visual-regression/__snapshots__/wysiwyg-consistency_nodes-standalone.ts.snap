// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WYSIWYG Comparison: actions 1`] = `
Object {
  "consistency": "100%",
  "divergence": 0,
}
`;

exports[`WYSIWYG Comparison: blockquote 1`] = `
Object {
  "consistency": "92.29%",
  "divergence": 0.07712418300653595,
}
`;

exports[`WYSIWYG Comparison: bullet list 1`] = `
Object {
  "consistency": "99.98%",
  "divergence": 0.0001513840830449827,
}
`;

exports[`WYSIWYG Comparison: codeblock 1`] = `
Object {
  "consistency": "97.87%",
  "divergence": 0.021318527410964384,
}
`;

exports[`WYSIWYG Comparison: date 1`] = `
Object {
  "consistency": "98.79%",
  "divergence": 0.012096774193548387,
}
`;

exports[`WYSIWYG Comparison: decisions 1`] = `
Object {
  "consistency": "100%",
  "divergence": 0,
}
`;

exports[`WYSIWYG Comparison: divider 1`] = `
Object {
  "consistency": "100%",
  "divergence": 0,
}
`;

exports[`WYSIWYG Comparison: emoji 1`] = `
Object {
  "consistency": "98.68%",
  "divergence": 0.01318785578747628,
}
`;

exports[`WYSIWYG Comparison: heading 1`] = `
Object {
  "consistency": "98.92%",
  "divergence": 0.010814203730272597,
}
`;

exports[`WYSIWYG Comparison: mention 1`] = `
Object {
  "consistency": "99.29%",
  "divergence": 0.007068311195445921,
}
`;

exports[`WYSIWYG Comparison: ordered list 1`] = `
Object {
  "consistency": "99.98%",
  "divergence": 0.0001513840830449827,
}
`;

exports[`WYSIWYG Comparison: panel 1`] = `
Object {
  "consistency": "100%",
  "divergence": 0,
}
`;

exports[`WYSIWYG Comparison: paragraph 1`] = `
Object {
  "consistency": "92.83%",
  "divergence": 0.07167755991285403,
}
`;

exports[`WYSIWYG Comparison: status 1`] = `
Object {
  "consistency": "99.33%",
  "divergence": 0.006688804554079697,
}
`;

exports[`WYSIWYG Comparison: table 1`] = `
Object {
  "consistency": "95.01%",
  "divergence": 0.04986787683823529,
}
`;
